File: Joon/JS/baekjoon/17298_오큰수.js
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n");

const n = +input[0];
const stack = [];

// [3,5,2,7]
const numbers = input[1].split(" ").map(Number);

// [-1,-1,-1,-1]
let result = new Array(n).fill(-1);

for (let i = 0; i < n; i++) {
  while (stack.length && numbers[stack[stack.length - 1]] < numbers[i]) {
    result[stack.pop()] = numbers[i];
  }
  stack.push(i);
}

console.log(result.join(" "));

File: Joon/Python/baekjoon/17298_오큰수.py
from collections import deque
n = int(input())
numbers = list(map(int,input().split()));
stack = deque()
result = [-1 for _ in range(n)]
for i in range(n):
    while stack and (numbers[stack[-1]] < numbers[i]):
        idx = stack.pop()
        result[idx] = numbers[i];
    stack.append(i)

while stack:
    result[stack.pop()] = -1

print(*result)
